math_opt_benchmark/mst/mst.cc

```diff
@@ -0,0 +1,119 @@
+using ::operations_research::MPSolver;
+using ::operations_research::MPVariable;
+
+/**
+ * @param problem_type MPSolver constant specifying which solver to use
+ * @param problem Graph specification with edges, weights, and number of
+ * vertices
+ */
+MSTSolver::MSTSolver(const MPSolver::OptimizationProblemType problem_type,
+                     const MSTProblem &problem)
+    : solver_("MST_solver", problem_type) {
+  solver_.MutableObjective()->SetMinimization();
+  for (int i = 0; i < problem.n; ++i) {
+    x_vars_.push_back(
+        std::vector<operations_research::MPVariable *>(problem.n, nullptr));
+  }
+  for (int i = 0; i < problem.n; ++i) {
+    for (int j = i; j < problem.n; ++j) {
+      MPVariable *var =
+          solver_.MakeVar(0.0, 1.0, problem.integer, absl::StrCat("x", i, j));
+      x_vars_[i][j] = x_vars_[j][i] = var;
+      UpdateObjective(i, j, problem.weights[i][j]);
+    }
+  }
+  for (int i = 0; i < problem.n; ++i) {
+    MPConstraint *c = solver_.MakeRowConstraint(0, 0);
+    c->SetCoefficient(x_vars_[i][i], 1);
+  }
+  MPConstraint *c_eq = solver_.MakeRowConstraint(problem.n - 1, problem.n - 1);
+  for (int i = 0; i < problem.n; ++i) {
+    for (int j = i + 1; j < problem.n; ++j) {
+      c_eq->SetCoefficient(x_vars_[i][j], 1);
+    }
+  }
+}
+
+void debug_solve(const MSTSolution &result) {
+  for (int i = 0; i < result.x_values.size(); i++) {
+    for (int j = 0; j < result.x_values[i].size(); j++) {
+      printf("[D] x[%i][%i] = %.7f\n", i, j, result.x_values[i][j]);
+    }
+  }
+}
+
+/**
+ * Calls the MPSolver solve routine and stores the result in a solution object
+ * @return Solution containing objective value and variable values
+ */
+MSTSolution MSTSolver::Solve() {
+  MPSolver::ResultStatus status = solver_.Solve();
+  CHECK_EQ(status, MPSolver::OPTIMAL);
+  MSTSolution result;
+  result.objective_value = solver_.Objective().Value();
+  result.x_values = std::vector<std::vector<double>>(
+      x_vars_.size(), std::vector<double>(x_vars_.size(), 0));
+  for (int i = 0; i < x_vars_.size(); i++) {
+    for (int j = 0; j < x_vars_.size(); j++) {
+      result.x_values[i][j] = (x_vars_[i][j]->solution_value());
+    }
+  }
+  return result;
+}
+
+/**
+ * Verifies input and updates the coefficient for a variable
+ * @param v1 Vertex 1
+ * @param v2 Vertex 2
+ * @param value Variable coefficient
+ */
+void MSTSolver::UpdateObjective(int v1, int v2, double value) {
+  CHECK_GE(v1, 0);
+  CHECK_GE(v2, 0);
+  CHECK_LT(v1, x_vars_.size());
+  CHECK_LT(v2, x_vars_[v1].size());
+  solver_.MutableObjective()->SetCoefficient(x_vars_[v1][v2], value);
+}
+
+/**
+ *
+ * @param problem Graph specification with edges, weights, and number of
+ * vertices
+ * @param invalid List of lists of vertices that violate the MST constraint
+ */
+void MSTSolver::AddConstraints(const MSTProblem &problem,
+                               std::vector<std::vector<int>> invalid) {
+  static constexpr int max_new_constraints = 25;
```

Comment:
just an optimization, but it is probably better to sort the constraints by largest violation. Also, smaller constraints (which are sparser) are also good.